File: apps/worker/src/cluster/cluster.ts
import type { NormalizedEntry } from '../ingest/types';
import { tokenize, jaccardSimilarity } from '../normalize/title_tokens';
import {
  findRecentStories,
  createStory,
  updateStoryLastUpdate,
} from '../db/stories_repo';
import { attachItem } from '../db/story_items_repo';

/** Look back 24 hours for candidate stories. */
const CLUSTER_WINDOW_MS = 24 * 60 * 60 * 1000;

/** Minimum Jaccard similarity required to attach to an existing story. */
const SIMILARITY_THRESHOLD = 0.25;

export interface ClusterCounters {
  storiesNew: number;
  storiesUpdated: number;
}

type ClusterItem = Pick<NormalizedEntry, 'itemKey' | 'titleHe' | 'publishedAt'>;

/**
 * Cluster a list of newly-ingested items into stories.
 *
 * Algorithm (per item):
 *  1. Tokenize item title.
 *  2. Scan candidate stories (from last 24 h) for highest Jaccard similarity.
 *  3. If similarity ≥ THRESHOLD → attach item; bump story.last_update_at.
 *  4. Otherwise → create a new story, attach item.
 *  5. In-memory candidate map is updated after each attach so that items
 *     within the same run can match a story created by an earlier item.
 *
 * All DB writes are INSERT OR IGNORE / idempotent, so re-running is safe.
 */
export async function clusterNewItems(
  db: D1Database,
  items: ClusterItem[],
): Promise<ClusterCounters> {
  if (items.length === 0) return { storiesNew: 0, storiesUpdated: 0 };

  const candidates = await findRecentStories(db, CLUSTER_WINDOW_MS);

  // In-memory map: storyId → token set (updated as items are clustered).
  const candidateTokens = new Map<string, Set<string>>(
    candidates.map(c => [c.storyId, tokenize(c.titleHe)]),
  );

  const counters: ClusterCounters = { storiesNew: 0, storiesUpdated: 0 };
  const now = new Date().toISOString();

  for (const item of items) {
    const itemTokens = tokenize(item.titleHe);

    // Find best-matching story above threshold.
    let bestStoryId: string | null = null;
    let bestScore = SIMILARITY_THRESHOLD; // must exceed (not just equal)

    for (const [storyId, storyTokens] of candidateTokens) {
      const score = jaccardSimilarity(itemTokens, storyTokens);
      if (score > bestScore) {
        bestScore = score;
        bestStoryId = storyId;
      }
    }

    if (bestStoryId !== null) {
      const attached = await attachItem(db, bestStoryId, item.itemKey, now);
      if (attached) {
        await updateStoryLastUpdate(db, bestStoryId, now);
        // Expand candidate tokens so subsequent items in this run can match.
        const existing = candidateTokens.get(bestStoryId);
        if (existing !== undefined) {
          for (const t of itemTokens) existing.add(t);
        }
        counters.storiesUpdated++;
      }
    } else {
      const newStoryId = crypto.randomUUID();
      const startAt = item.publishedAt ?? now;
      await createStory(db, newStoryId, startAt);
      await attachItem(db, newStoryId, item.itemKey, now);
      // Register new story so items later in this run can match it.
      candidateTokens.set(newStoryId, new Set(itemTokens));
      counters.storiesNew++;
    }
  }

  return counters;
}

File: apps/local-engine/cluster/cluster.py
"""cluster/cluster.py — Jaccard-based story clustering.

Exact port of apps/worker/src/cluster/cluster.ts.
Algorithm, thresholds, and window MUST remain identical so that the local
engine produces equivalent story groupings to the Cloudflare Worker.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import UTC, datetime
from uuid import uuid4

import aiosqlite

from cluster.tokens import jaccard_similarity, tokenize
from db.repos.stories_repo import (
    create_story,
    find_recent_stories,
    update_story_last_update,
)
from db.repos.story_items_repo import attach_item

# Look back 24 hours for candidate stories.
CLUSTER_WINDOW_SEC: int = 24 * 60 * 60

# Minimum Jaccard similarity required to attach to an existing story.
# Must EXCEED (strict >) the threshold — identical to TS `if (score > bestScore)`.
SIMILARITY_THRESHOLD: float = 0.25


@dataclass
class ClusterItem:
    """Minimal item representation needed for clustering."""

    item_key: str
    title_he: str
    published_at: str | None = None


@dataclass
class ClusterCounters:
    stories_new: int = 0
    stories_updated: int = 0


def _now_iso() -> str:
    return datetime.now(UTC).strftime("%Y-%m-%dT%H:%M:%S.%f")[:-3] + "Z"


async def cluster_new_items(
    db: aiosqlite.Connection,
    items: list[ClusterItem],
) -> ClusterCounters:
    """Cluster a list of newly-ingested items into stories.

    Algorithm (per item):
      1. Tokenize item title.
      2. Scan candidate stories (from last 24 h) for highest Jaccard similarity.
      3. If similarity > THRESHOLD → attach item; bump story.last_update_at.
      4. Otherwise → create a new story, attach item.
      5. In-memory candidate map is updated after each attach so that items
         within the same run can match a story created by an earlier item.

    All DB writes are INSERT OR IGNORE / idempotent, so re-running is safe.
    Commits once at the end of the batch.
    """
    if not items:
        return ClusterCounters()

    candidates = await find_recent_stories(db, CLUSTER_WINDOW_SEC)

    # In-memory map: story_id → token set (expanded as items are clustered).
    candidate_tokens: dict[str, frozenset[str]] = {
        c.story_id: tokenize(c.title_he) for c in candidates
    }

    counters = ClusterCounters()
    now = _now_iso()

    for item in items:
        item_tokens = tokenize(item.title_he)

        # Find best-matching story above threshold.
        best_story_id: str | None = None
        best_score = SIMILARITY_THRESHOLD  # must strictly exceed

        for story_id, story_tokens in candidate_tokens.items():
            score = jaccard_similarity(item_tokens, story_tokens)
            if score > best_score:
                best_score = score
                best_story_id = story_id

        if best_story_id is not None:
            attached = await attach_item(db, best_story_id, item.item_key, now)
            if attached:
                await update_story_last_update(db, best_story_id, now)
                # Expand candidate tokens so subsequent items can match.
                existing = candidate_tokens[best_story_id]
                candidate_tokens[best_story_id] = existing | item_tokens
                counters.stories_updated += 1
        else:
            new_story_id = str(uuid4())
            start_at = item.published_at or now
            await create_story(db, new_story_id, start_at)
            await attach_item(db, new_story_id, item.item_key, now)
            # Register new story so later items in this run can match it.
            candidate_tokens[new_story_id] = item_tokens
            counters.stories_new += 1

    await db.commit()
    return counters
